### PR TITLE
Azure SQL Code Modifications and Integration Tests Added

### PR DESCRIPTION
**Pull request by kedarghule**

### CODE REVIEW

File: cartography/intel/azure/sql.py [hunk 1]
```diff
@@ -277,267 +291,237 @@ def load_server_details(neo4j_session, credentials, subscription_id, details, up
...
+        neo4j_session: neo4j.Session, dns_aliases: List[Dict], update_tag: int
+) -> None:
     """
     Ingest the DNS Alias details into neo4j.
     """
     ingest_dns_aliases = """
-    MERGE (alias:AzureServerDNSAlias{id: {DNSAliasId}})
+    UNWIND {dns_aliases_list} as dns_alias
+    MERGE (alias:AzureServerDNSAlias{id: dns_alias.id})
     ON CREATE SET alias.firstseen = timestamp(), alias.lastupdated = {azure_update_tag}
```
Comment (mpurusottamc) [inline cartography/intel/azure/sql.py:303]: lastupdated should not be in the create set. it should be in update set.

File: cartography/intel/azure/sql.py [hunk 2]
```diff
@@ -277,267 +291,237 @@ def load_server_details(neo4j_session, credentials, subscription_id, details, up
...
+        neo4j_session: neo4j.Session, ad_admins: List[Dict], update_tag: int
+) -> None:
     """
     Ingest the Server AD Administrators details into neo4j.
     """
     ingest_ad_admins = """
-    MERGE (a:AzureServerADAdministrator{id: {AdAdminId}})
+    UNWIND {ad_admins_list} as ad_admins
+    MERGE (a:AzureServerADAdministrator{id: ad_admins.id})
     ON CREATE SET a.firstseen = timestamp(), a.lastupdated = {azure_update_tag}
```
Comment (mpurusottamc) [inline cartography/intel/azure/sql.py:330]: lastupdated should not be in the create set. it should be in update set.

File: cartography/intel/azure/sql.py [hunk 3]
```diff
@@ -277,267 +291,237 @@ def load_server_details(neo4j_session, credentials, subscription_id, details, up
...
-def _load_server_ad_admins(neo4j_session, ad_admins, update_tag):
+def _load_server_ad_admins(
+        neo4j_session: neo4j.Session, ad_admins: List[Dict], update_tag: int
+) -> None:
     """
     Ingest the Server AD Administrators details into neo4j.
     """
     ingest_ad_admins = """
-    MERGE (a:AzureServerADAdministrator{id: {AdAdminId}})
+    UNWIND {ad_admins_list} as ad_admins
```
Comment (mpurusottamc) [inline cartography/intel/azure/sql.py:328]: let's rename ad_admins to ad_admin, since it's a single object

File: cartography/intel/azure/sql.py [hunk 4]
```diff
@@ -277,267 +291,237 @@ def load_server_details(neo4j_session, credentials, subscription_id, details, up
...
+        neo4j_session: neo4j.Session, recoverable_databases: List[Dict], update_tag: int
+) -> None:
     """
     Ingest the recoverable database details into neo4j.
     """
     ingest_recoverable_databases = """
-    MERGE (rd:AzureRecoverableDatabase{id: {DatabaseId}})
+    UNWIND {recoverable_databases_list} as rec_dbs
+    MERGE (rd:AzureRecoverableDatabase{id: rec_dbs.id})
     ON CREATE SET rd.firstseen = timestamp(), rd.lastupdated = {azure_update_tag}
```
Comment (mpurusottamc) [inline cartography/intel/azure/sql.py:358]: lastupdated should not be in the create set. it should be in update set.

File: cartography/intel/azure/sql.py [hunk 5]
```diff
@@ -277,267 +291,237 @@ def load_server_details(neo4j_session, credentials, subscription_id, details, up
...
-def _load_recoverable_databases(neo4j_session, recoverable_databases, update_tag):
+def _load_recoverable_databases(
+        neo4j_session: neo4j.Session, recoverable_databases: List[Dict], update_tag: int
+) -> None:
     """
     Ingest the recoverable database details into neo4j.
     """
     ingest_recoverable_databases = """
-    MERGE (rd:AzureRecoverableDatabase{id: {DatabaseId}})
+    UNWIND {recoverable_databases_list} as rec_dbs
```
Comment (mpurusottamc) [inline cartography/intel/azure/sql.py:356]: let's rename rec_dbs to rec_db, since it's a single object

File: cartography/intel/azure/sql.py [hunk 6]
```diff
@@ -277,267 +291,237 @@ def load_server_details(neo4j_session, credentials, subscription_id, details, up
...
-def _load_restorable_dropped_databases(neo4j_session, restorable_dropped_databases, update_tag):
+def _load_restorable_dropped_databases(
+        neo4j_session: neo4j.Session, restorable_dropped_databases: List[Dict], update_tag: int
+) -> None:
     """
     Ingest the restorable dropped database details into neo4j.
     """
     ingest_restorable_dropped_databases = """
-    MERGE (rdd:AzureRestorableDroppedDatabase{id: {DatabaseId}})
+    UNWIND {restorable_dropped_databases_list} as res_dropped_dbs
```
Comment (mpurusottamc) [inline cartography/intel/azure/sql.py:385]: rename res_dropped_dbs to res_dropped_db

File: cartography/intel/azure/sql.py [hunk 7]
```diff
@@ -277,267 +291,237 @@ def load_server_details(neo4j_session, credentials, subscription_id, details, up
...
+    MERGE (rdd:AzureRestorableDroppedDatabase{id: res_dropped_dbs.id})
     ON CREATE SET rdd.firstseen = timestamp(), rdd.lastupdated = {azure_update_tag}
```
Comment (mpurusottamc) [inline cartography/intel/azure/sql.py:387]: lastupdated should not be in the create set. it should be in update set.

File: cartography/intel/azure/sql.py [hunk 8]
```diff
@@ -277,267 +291,237 @@ def load_server_details(neo4j_session, credentials, subscription_id, details, up
...
+        neo4j_session: neo4j.Session, failover_groups: List[Dict], update_tag: int
+) -> None:
     """
     Ingest the failover groups details into neo4j.
     """
     ingest_failover_groups = """
-    MERGE (f:AzureFailoverGroup{id: {GroupId}})
+    UNWIND {failover_groups_list} as fg
+    MERGE (f:AzureFailoverGroup{id: fg.id})
     ON CREATE SET f.firstseen = timestamp(), f.lastupdated = {azure_update_tag}
```
Comment (mpurusottamc) [inline cartography/intel/azure/sql.py:421]: lastupdated should not be in the create set. it should be in update set.

File: cartography/intel/azure/sql.py [hunk 9]
```diff
@@ -277,267 +291,237 @@ def load_server_details(neo4j_session, credentials, subscription_id, details, up
...
+        neo4j_session: neo4j.Session, elastic_pools: List[Dict], update_tag: int
+) -> None:
     """
     Ingest the elastic pool details into neo4j.
     """
     ingest_elastic_pools = """
-    MERGE (e:AzureElasticPool{id: {PoolId}})
+    UNWIND {elastic_pools_list} as ep
+    MERGE (e:AzureElasticPool{id: ep.id})
     ON CREATE SET e.firstseen = timestamp(), e.lastupdated = {azure_update_tag}
```
Comment (mpurusottamc) [inline cartography/intel/azure/sql.py:450]: lastupdated should not be in the create set. it should be in update set.

File: cartography/intel/azure/sql.py [hunk 10]
```diff
@@ -277,267 +291,237 @@ def load_server_details(neo4j_session, credentials, subscription_id, details, up
...
+        neo4j_session: neo4j.Session, databases: List[Dict], update_tag: int
+) -> None:
     """
     Ingest the database details into neo4j.
     """
     ingest_databases = """
-    MERGE (d:AzureDatabase{id: {Id}})
+    UNWIND {databases_list} as az_database
+    MERGE (d:AzureSQLDatabase{id: az_database.id})
     ON CREATE SET d.firstseen = timestamp(), d.lastupdated = {azure_update_tag}
```
Comment (mpurusottamc) [inline cartography/intel/azure/sql.py:483]: lastupdated should not be in the create set. it should be in update set.

File: cartography/intel/azure/sql.py [hunk 11]
```diff
@@ -277,267 +291,237 @@ def load_server_details(neo4j_session, credentials, subscription_id, details, up
...
-    d.secondarylocation = {SecondaryLocation},
-    d.elasticpoolid = {ElasticPoolId},
-    d.collation = {Collation},
-    d.failovergroupid = {FailoverGroupId},
-    d.restorabledroppeddbid = {RDDId},
-    d.recoverabledbid = {RecoverableDbId}
-    WITH d
-    MATCH (s:AzureServer{id: {ServerId}})
+    SET d.name = az_database.name,
+    d.location = az_database.location,
```
Comment (mpurusottamc) [inline cartography/intel/azure/sql.py:485]: location cannot be changed after a resource is created. so, let's move it to the create set for all resources in sql, storage, etc.

File: cartography/intel/azure/sql.py [hunk 12]
```diff
@@ -647,169 +633,148 @@ def load_database_details(neo4j_session, details, update_tag):
...
+        neo4j_session: neo4j.Session, replication_links: List[Dict], update_tag: int,
+) -> None:
     """
     Ingest replication links into neo4j.
     """
     ingest_replication_links = """
-    MERGE (rl:AzureReplicationLink{id: {LinkId}})
+    UNWIND {replication_links_list} as replication_link
+    MERGE (rl:AzureReplicationLink{id: replication_link.id})
     ON CREATE SET rl.firstseen = timestamp(), rl.lastupdated = {azure_update_tag}
```
Comment (mpurusottamc) [inline cartography/intel/azure/sql.py:645]: same as above. move lastupdated to update set. move location to create set

File: cartography/intel/azure/sql.py [hunk 13]
```diff
@@ -647,169 +633,148 @@ def load_database_details(neo4j_session, details, update_tag):
...
+        neo4j_session: neo4j.Session, threat_detection_policies: List[Dict], update_tag: int,
+) -> None:
     """
     Ingest threat detection policy into neo4j.
     """
     ingest_threat_detection_policies = """
-    MERGE (policy:AzureDatabaseThreatDetectionPolicy{id: {PolicyId}})
+    UNWIND {threat_detection_policies_list} as tdp
+    MERGE (policy:AzureDatabaseThreatDetectionPolicy{id: tdp.id})
     ON CREATE SET policy.firstseen = timestamp(), policy.lastupdated = {azure_update_tag}
```
Comment (mpurusottamc) [inline cartography/intel/azure/sql.py:682]: same as above

File: cartography/intel/azure/sql.py [hunk 14]
```diff
@@ -647,169 +633,148 @@ def load_database_details(neo4j_session, details, update_tag):
...
+        neo4j_session: neo4j.Session, restore_points: List[Dict], update_tag: int,
+) -> None:
     """
     Ingest restore points into neo4j.
     """
     ingest_restore_points = """
-    MERGE (point:AzureRestorePoint{id: {PointId}})
+    UNWIND {restore_points_list} as rp
+    MERGE (point:AzureRestorePoint{id: rp.id})
     ON CREATE SET point.firstseen = timestamp(), point.lastupdated = {azure_update_tag}
```
Comment (mpurusottamc) [inline cartography/intel/azure/sql.py:717]: same as above

File: cartography/intel/azure/sql.py [hunk 15]
```diff
@@ -647,169 +633,148 @@ def load_database_details(neo4j_session, details, update_tag):
...
+        neo4j_session: neo4j.Session, encryptions_list: List[Dict], update_tag: int,
+) -> None:
     """
     Ingest transparent data encryptions into neo4j.
     """
     ingest_data_encryptions = """
-    MERGE (tae:AzureTransparentDataEncryption{id: {TAEId}})
+    UNWIND {transparent_data_encryptions_list} as e
+    MERGE (tae:AzureTransparentDataEncryption{id: e.id})
     ON CREATE SET tae.firstseen = timestamp(), tae.lastupdated = {azure_update_tag}
```
Comment (mpurusottamc) [inline cartography/intel/azure/sql.py:747]: same as above. lastupdated & location needs movement between sets

File: cartography/intel/azure/sql.py [hunk 16]
```diff
@@ -647,169 +633,148 @@ def load_database_details(neo4j_session, details, update_tag):
...
+        azure_update_tag=update_tag,
+    )
 
 
 @timeit
-def cleanup_azure_sql_servers(neo4j_session, subscription_id, common_job_parameters):
+def cleanup_azure_sql_servers(
+        neo4j_session: neo4j.Session, subscription_id: str, common_job_parameters: Dict
+) -> None:
     common_job_parameters['AZURE_SUBSCRIPTION_ID'] = subscription_id
```
Comment (mpurusottamc) [inline cartography/intel/azure/sql.py:769]: We don't need this assignment. This would be handled in __init__.py file.

**Review by mpurusottamc — CHANGES_REQUESTED** [review]
There are few changes which needs to be incorporated. Let me know if you have any questions.

**Comment by mpurusottamc** [comment]
update setup.py to include dependencies

File: cartography/intel/azure/sql.py [hunk 17]
```diff
@@ -37,51 +46,53 @@ def get_server_list(credentials, subscription_id):
...
     Ingest the server details into neo4j.
     """
     ingest_server = """
-    MERGE (s:AzureServer{id: {ServerId}})
+    UNWIND {server_list} as server
+    MERGE (s:AzureSQLServer{id: server.id})
     ON CREATE SET s.firstseen = timestamp(),
-    s.id = {ServerId}, s.name = {Name},
-    s.resourcegroup = {ResourceGroup}, s.location = {Location}
+    s.name = server.name, s.resourcegroup = server.resourceGroup,
```
[the remark this answers is not in the document]
Comment (kedarghule) [reply cartography/intel/azure/sql.py:60]: Done

File: cartography/intel/azure/sql.py (hunk 1 above)
Comment (kedarghule) [reply cartography/intel/azure/sql.py:303]: Done

File: cartography/intel/azure/sql.py (hunk 2 above)
Comment (kedarghule) [reply cartography/intel/azure/sql.py:330]: done

File: cartography/intel/azure/sql.py (hunk 3 above)
Comment (kedarghule) [reply cartography/intel/azure/sql.py:328]: done

File: cartography/intel/azure/sql.py (hunk 5 above)
Comment (kedarghule) [reply cartography/intel/azure/sql.py:356]: Done

File: cartography/intel/azure/sql.py (hunk 4 above)
Comment (kedarghule) [reply cartography/intel/azure/sql.py:358]: Done

File: cartography/intel/azure/sql.py (hunk 6 above)
Comment (kedarghule) [reply cartography/intel/azure/sql.py:385]: Done

File: cartography/intel/azure/sql.py (hunk 7 above)
Comment (kedarghule) [reply cartography/intel/azure/sql.py:387]: done

File: cartography/intel/azure/sql.py (hunk 8 above)
Comment (kedarghule) [reply cartography/intel/azure/sql.py:421]: done

File: cartography/intel/azure/sql.py (hunk 10 above)
Comment (kedarghule) [reply cartography/intel/azure/sql.py:483]: done

File: cartography/intel/azure/sql.py (hunk 9 above)
Comment (kedarghule) [reply cartography/intel/azure/sql.py:450]: done

File: cartography/intel/azure/sql.py (hunk 11 above)
Comment (kedarghule) [reply cartography/intel/azure/sql.py:485]: Done

File: cartography/intel/azure/sql.py (hunk 12 above)
Comment (kedarghule) [reply cartography/intel/azure/sql.py:645]: done

File: cartography/intel/azure/sql.py (hunk 13 above)
Comment (kedarghule) [reply cartography/intel/azure/sql.py:682]: done

File: cartography/intel/azure/sql.py (hunk 14 above)
Comment (kedarghule) [reply cartography/intel/azure/sql.py:717]: done

File: cartography/intel/azure/sql.py (hunk 15 above)
Comment (kedarghule) [reply cartography/intel/azure/sql.py:747]: done

File: cartography/intel/azure/sql.py (hunk 16 above)
Comment (kedarghule) [reply cartography/intel/azure/sql.py:769]: Done

File: cartography/intel/azure/sql.py [hunk 18]
```diff
@@ -95,41 +106,41 @@ def get_server_details(credentials, subscription_id, server_list):
 
 
 @timeit
```
[the remark this answers is not in the document]
Comment (kedarghule) [reply cartography/intel/azure/sql.py:109]: I did it using Pycharm but it hardly made any changes here.

File: cartography/intel/azure/sql.py [hunk 19]
```diff
@@ -95,41 +106,41 @@ def get_server_details(credentials, subscription_id, server_list):
...
-def get_recoverable_databases(credentials, subscription_id, server):
+def get_recoverable_databases(credentials: Credentials, subscription_id: str, server: Dict) -> List[Dict]:
     """
-    Returns details of Recoverable databases in a server.
+    Returns details of the Recoverable databases in a server.
     """
     try:
         client = get_client(credentials, subscription_id)
```
[the remark this answers is not in the document]
Comment (kedarghule) [reply cartography/intel/azure/sql.py:150]: done

File: tests/data/azure/sql.py [hunk 20]
```diff
@@ -0,0 +1,325 @@
+DESCRIBE_SERVERS = [
```
[the remark this answers is not in the document]
Comment (kedarghule) [reply tests/data/azure/sql.py:1]: PyCharm isnt reducing lines for strings. If I do it then it marks it as an error

File: tests/integration/cartography/intel/azure/test_sql.py [hunk 21]
```diff
@@ -0,0 +1,773 @@
+from cartography.intel.azure import sql
```
[the remark this answers is not in the document]
Comment (kedarghule) [reply tests/integration/cartography/intel/azure/test_sql.py:1]: PyCharm isnt reducing lines for strings. If I do it then it marks it as an error

**Comment by kedarghule** [comment]
@mpurusottamc made necessary changes. lemme know

**Review by mpurusottamc — APPROVED** [review]
Looks good. There are few additional spaces in the cypher queries. we can fix them when we open the Pull Request :)

**Pull request closed without merging**